test/unit/Main.hs

```diff
@@ -56,3 +58,70 @@ matchingTests = testGroup "Matching"
     tryQuery expectedVal query input = assertBool "Check for match" $
       matchingString input query == expectedVal
 
+simplificationTests =
+  testGroup
+    "Expression simplification"
+    [ testGroup
+        "Return true if the expression can be simplified to False"
+        [ testGroup
+            "Effect of constant literals"
+            [ testCase
+                "False input via single literal"
+                $ expectTrue BFalse
+            , testCase
+                "True input via composed literals"
+                $ expectFalse BTrue
+            , testCase
+                "False input via composed literals"
+                $ expectTrue
+                $ BOr BFalse BFalse
+            , testCase
+                "True input via composed literals"
+                $ expectFalse
+                $ BOr BFalse BTrue
+            , testCase
+                "Constant OR'd with False"
+                $ expectFalse
+                $ BOr BFalse (BConst (Positive "foo"))
+            , testCase
+                "Constant OR'd with True"
+                $ expectFalse
+                $ BOr (BConst (Positive "foo")) BTrue
+            , testCase
+                "Constant AND'd with False"
+                $ expectTrue
+                $ BAnd BFalse (BConst (Positive "foo"))
+            , testCase
+                "Constant AND'd with True"
+                $ expectFalse
+                $ BAnd (BConst (Positive "foo")) BTrue
+            , testCase
+                "Nested Constants AND'd with False within OR"
+                $ expectTrue
+                $ BOr
+                  (BAnd BFalse (BConst (Positive "foo")))
+                  (BAnd (BConst (Positive "bar")) BFalse)
+            , testCase
+                "Deeply nested Constants AND'd with False within OR with multiple negations"
+                $ expectTrue
+                $ BOr
+                  (BAnd (BNot BTrue) (BNot (BNot (BNot (BConst (Positive "foo"))))))
+                  (BAnd (BConst (Positive "bar")) (BNot (BNot BFalse)))
+            ]
+        , testGroup
+            "Effect of contradicting named constants"
+            [ testCase
+                "via NOT operator"
+                $ expectTrue
+                $ BAnd (BNot (BConst (Positive "foo"))) (BConst (Positive "foo"))
+            , testCase
+                "via signedness"
+                $ expectTrue
+                $ BAnd (BConst (Positive "foo")) (BConst (Negative "foo"))
+            ]
+        ]
+    ]
+  where
+    expectTrue, expectFalse :: BoolExpr String -> Assertion
+    expectTrue = assertBool "Should have returned true" . cannotBeTrue
+    expectFalse = assertBool "Should have returned false" . not . cannotBeTrue
```

Comment:
I'm troubled by the naming convention of `expectTrue` and `expectFalse`, even if I understand because it is the test of `cannotBeTrue` but I still find it confusing. At least renaming `expectTrue` to `expectCannotBeTrue` and `expectFalse` to `canBeTrue`.

Reply:
Renamed.